File: system/sensord/sensors/lsm6ds3_gyro.cc
#include "lsm6ds3_gyro.h"

#include <cassert>
#include <cmath>
#include <cstring>

#include "common/swaglog.h"
#include "common/timing.h"
#include "common/util.h"

#define DEG2RAD(x) ((x) * M_PI / 180.0)

LSM6DS3_Gyro::LSM6DS3_Gyro(I2CBus *bus, int gpio_nr, bool shared_gpio) :
  I2CSensor(bus, gpio_nr, shared_gpio) {}

void LSM6DS3_Gyro::wait_for_data_ready() {
  uint8_t drdy = 0;
  uint8_t buffer[6];

  do {
    read_register(LSM6DS3_GYRO_I2C_REG_STAT_REG, &drdy, sizeof(drdy));
    drdy &= LSM6DS3_GYRO_DRDY_GDA;
  } while (drdy == 0);

  read_register(LSM6DS3_GYRO_I2C_REG_OUTX_L_G, buffer, sizeof(buffer));
}

void LSM6DS3_Gyro::read_and_avg_data(float* out_buf) {
  uint8_t drdy = 0;
  uint8_t buffer[6];

  for (int i = 0; i < 5; i++) {
    do {
      read_register(LSM6DS3_GYRO_I2C_REG_STAT_REG, &drdy, sizeof(drdy));
      drdy &= LSM6DS3_GYRO_DRDY_GDA;
    } while (drdy == 0);

    int len = read_register(LSM6DS3_GYRO_I2C_REG_OUTX_L_G, buffer, sizeof(buffer));
    assert(len == sizeof(buffer));

    for (int j = 0; j < 3; j++) {
      out_buf[j] += (float)read_16_bit(buffer[j*2], buffer[j*2+1]) * 70.0f;
    }
  }

  // calculate the mg average values
  for (int i = 0; i < 3; i++) {
    out_buf[i] /= 5.0f;
  }
}

int LSM6DS3_Gyro::self_test(int test_type) {
  float val_st_off[3] = {0};
  float val_st_on[3] = {0};
  float test_val[3] = {0};

  // prepare sensor for self-test

  // full scale: 2000dps, ODR: 208Hz
  int ret = set_register(LSM6DS3_GYRO_I2C_REG_CTRL2_G, LSM6DS3_GYRO_ODR_208HZ | LSM6DS3_GYRO_FS_2000dps);
  if (ret < 0) {
    return ret;
  }

  // wait for stable output, and discard first values
  util::sleep_for(150);
  wait_for_data_ready();
  read_and_avg_data(val_st_off);

  // enable Self Test positive (or negative)
  ret = set_register(LSM6DS3_GYRO_I2C_REG_CTRL5_C, test_type);
  if (ret < 0) {
    return ret;
  }

  // wait for stable output, and discard first values
  util::sleep_for(50);
  wait_for_data_ready();
  read_and_avg_data(val_st_on);

  // disable sensor
  ret = set_register(LSM6DS3_GYRO_I2C_REG_CTRL2_G, 0);
  if (ret < 0) {
    return ret;
  }

  // disable self test
  ret = set_register(LSM6DS3_GYRO_I2C_REG_CTRL5_C, 0);
  if (ret < 0) {
    return ret;
  }

  // calculate the mg values for self test
  for (int i = 0; i < 3; i++) {
    test_val[i] = fabs(val_st_on[i] - val_st_off[i]);
  }

  // verify test result
  for (int i = 0; i < 3; i++) {
    if ((LSM6DS3_GYRO_MIN_ST_LIMIT_mdps > test_val[i]) ||
        (test_val[i] > LSM6DS3_GYRO_MAX_ST_LIMIT_mdps)) {
      return -1;
    }
  }

  return ret;
}

int LSM6DS3_Gyro::init() {
  uint8_t value = 0;
  bool do_self_test = false;

  const char* env_lsm_selftest =env_lsm_selftest = std::getenv("LSM_SELF_TEST");
  if (env_lsm_selftest != nullptr && strncmp(env_lsm_selftest, "1", 1) == 0) {
    do_self_test = true;
  }

  int ret = verify_chip_id(LSM6DS3_GYRO_I2C_REG_ID, {LSM6DS3_GYRO_CHIP_ID, LSM6DS3TRC_GYRO_CHIP_ID});
  if (ret == -1) return -1;

  if (ret == LSM6DS3TRC_GYRO_CHIP_ID) {
    source = cereal::SensorEventData::SensorSource::LSM6DS3TRC;
  }

  ret = init_gpio();
  if (ret < 0) {
    goto fail;
  }

  ret = self_test(LSM6DS3_GYRO_POSITIVE_TEST);
  if (ret < 0 ) {
    LOGE("LSM6DS3 gyro positive self-test failed!");
    if (do_self_test) goto fail;
  }

  ret = self_test(LSM6DS3_GYRO_NEGATIVE_TEST);
  if (ret < 0) {
    LOGE("LSM6DS3 gyro negative self-test failed!");
    if (do_self_test) goto fail;
  }

  // TODO: set scale. Default is +- 250 deg/s
  ret = set_register(LSM6DS3_GYRO_I2C_REG_CTRL2_G, LSM6DS3_GYRO_ODR_104HZ);
  if (ret < 0) {
    goto fail;
  }

  ret = set_register(LSM6DS3_GYRO_I2C_REG_DRDY_CFG, LSM6DS3_GYRO_DRDY_PULSE_MODE);
  if (ret < 0) {
    goto fail;
  }

  // enable data ready interrupt for gyro on INT1
  // (without resetting existing interrupts)
  ret = read_register(LSM6DS3_GYRO_I2C_REG_INT1_CTRL, &value, 1);
  if (ret < 0) {
    goto fail;
  }

  value |= LSM6DS3_GYRO_INT1_DRDY_G;
  ret = set_register(LSM6DS3_GYRO_I2C_REG_INT1_CTRL, value);

fail:
  return ret;
}

int LSM6DS3_Gyro::shutdown() {
  int ret = 0;

  // disable data ready interrupt for gyro on INT1
  uint8_t value = 0;
  ret = read_register(LSM6DS3_GYRO_I2C_REG_INT1_CTRL, &value, 1);
  if (ret < 0) {
    goto fail;
  }

  value &= ~(LSM6DS3_GYRO_INT1_DRDY_G);
  ret = set_register(LSM6DS3_GYRO_I2C_REG_INT1_CTRL, value);
  if (ret < 0) {
    LOGE("Could not disable lsm6ds3 gyroscope interrupt!");
    goto fail;
  }

  // enable power-down mode
  value = 0;
  ret = read_register(LSM6DS3_GYRO_I2C_REG_CTRL2_G, &value, 1);
  if (ret < 0) {
    goto fail;
  }

  value &= 0x0F;
  ret = set_register(LSM6DS3_GYRO_I2C_REG_CTRL2_G, value);
  if (ret < 0) {
    LOGE("Could not power-down lsm6ds3 gyroscope!");
    goto fail;
  }

fail:
  return ret;
}

bool LSM6DS3_Gyro::get_event(MessageBuilder &msg, uint64_t ts) {

  // INT1 shared with accel, check STATUS_REG who triggered
  uint8_t status_reg = 0;
  read_register(LSM6DS3_GYRO_I2C_REG_STAT_REG, &status_reg, sizeof(status_reg));
  if ((status_reg & LSM6DS3_GYRO_DRDY_GDA) == 0) {
    return false;
  }

  uint8_t buffer[6];
  int len = read_register(LSM6DS3_GYRO_I2C_REG_OUTX_L_G, buffer, sizeof(buffer));
  assert(len == sizeof(buffer));

  float scale = 8.75 / 1000.0;
  float x = DEG2RAD(read_16_bit(buffer[0], buffer[1]) * scale);
  float y = DEG2RAD(read_16_bit(buffer[2], buffer[3]) * scale);
  float z = DEG2RAD(read_16_bit(buffer[4], buffer[5]) * scale);

  auto event = msg.initEvent().initGyroscope();
  event.setSource(source);
  event.setVersion(2);
  event.setSensor(SENSOR_GYRO_UNCALIBRATED);
  event.setType(SENSOR_TYPE_GYROSCOPE_UNCALIBRATED);
  event.setTimestamp(ts);

  float xyz[] = {y, -x, z};
  auto svec = event.initGyroUncalibrated();
  svec.setV(xyz);
  svec.setStatus(true);

  return true;
}

File: system/sensord/sensors/lsm6ds3_gyro.py
import os
import math
import time

from cereal import log
from openpilot.system.sensord.sensors.i2c_sensor import Sensor

class LSM6DS3_Gyro(Sensor):
  LSM6DS3_GYRO_I2C_REG_DRDY_CFG  = 0x0B
  LSM6DS3_GYRO_I2C_REG_INT1_CTRL = 0x0D
  LSM6DS3_GYRO_I2C_REG_CTRL2_G   = 0x11
  LSM6DS3_GYRO_I2C_REG_CTRL5_C   = 0x14
  LSM6DS3_GYRO_I2C_REG_STAT_REG  = 0x1E
  LSM6DS3_GYRO_I2C_REG_OUTX_L_G  = 0x22

  LSM6DS3_GYRO_ODR_104HZ       = (0b0100 << 4)
  LSM6DS3_GYRO_INT1_DRDY_G     = 0b10
  LSM6DS3_GYRO_DRDY_GDA        = 0b10
  LSM6DS3_GYRO_DRDY_PULSE_MODE = (1 << 7)

  LSM6DS3_GYRO_ODR_208HZ       = (0b0101 << 4)
  LSM6DS3_GYRO_FS_2000dps      = (0b11 << 2)
  LSM6DS3_GYRO_POSITIVE_TEST   = (0b01 << 2)
  LSM6DS3_GYRO_NEGATIVE_TEST   = (0b11 << 2)
  LSM6DS3_GYRO_MIN_ST_LIMIT_mdps = 150000.0
  LSM6DS3_GYRO_MAX_ST_LIMIT_mdps = 700000.0

  @property
  def device_address(self) -> int:
    return 0x6A

  def init(self):
    chip_id = self.verify_chip_id(0x0F, [0x69, 0x6A])
    if chip_id == 0x6A:
      self.source = log.SensorEventData.SensorSource.lsm6ds3trc
    else:
      self.source = log.SensorEventData.SensorSource.lsm6ds3

    # self-test
    if "LSM_SELF_TEST" in os.environ:
      self.self_test(self.LSM6DS3_GYRO_POSITIVE_TEST)
      self.self_test(self.LSM6DS3_GYRO_NEGATIVE_TEST)

    # actual init
    self.writes((
      # TODO: set scale. Default is +- 250 deg/s
      (self.LSM6DS3_GYRO_I2C_REG_CTRL2_G, self.LSM6DS3_GYRO_ODR_104HZ),
      # Configure data ready signal to pulse mode
      (self.LSM6DS3_GYRO_I2C_REG_DRDY_CFG, self.LSM6DS3_GYRO_DRDY_PULSE_MODE),
    ))
    value = self.read(self.LSM6DS3_GYRO_I2C_REG_INT1_CTRL, 1)[0]
    value |= self.LSM6DS3_GYRO_INT1_DRDY_G
    self.write(self.LSM6DS3_GYRO_I2C_REG_INT1_CTRL, value)

  def get_event(self, ts: int | None = None) -> log.SensorEventData:
    assert ts is not None  # must come from the IRQ event

    # Check if gyroscope data is ready, since it's shared with accelerometer
    status_reg = self.read(self.LSM6DS3_GYRO_I2C_REG_STAT_REG, 1)[0]
    if not (status_reg & self.LSM6DS3_GYRO_DRDY_GDA):
      raise self.DataNotReady

    b = self.read(self.LSM6DS3_GYRO_I2C_REG_OUTX_L_G, 6)
    x = self.parse_16bit(b[0], b[1])
    y = self.parse_16bit(b[2], b[3])
    z = self.parse_16bit(b[4], b[5])
    scale = (8.75 / 1000.0) * (math.pi / 180.0)
    xyz = [y * scale, -x * scale, z * scale]

    event = log.SensorEventData.new_message()
    event.timestamp = ts
    event.version = 2
    event.sensor = 5  # SENSOR_GYRO_UNCALIBRATED
    event.type = 16   # SENSOR_TYPE_GYROSCOPE_UNCALIBRATED
    event.source = self.source
    g = event.init('gyroUncalibrated')
    g.v = xyz
    g.status = 1
    return event

  def shutdown(self) -> None:
    # Disable data ready interrupt on INT1
    value = self.read(self.LSM6DS3_GYRO_I2C_REG_INT1_CTRL, 1)[0]
    value &= ~self.LSM6DS3_GYRO_INT1_DRDY_G
    self.write(self.LSM6DS3_GYRO_I2C_REG_INT1_CTRL, value)

    # Power down by clearing ODR bits
    value = self.read(self.LSM6DS3_GYRO_I2C_REG_CTRL2_G, 1)[0]
    value &= 0x0F
    self.write(self.LSM6DS3_GYRO_I2C_REG_CTRL2_G, value)

  # *** self-test stuff ***
  def _wait_for_data_ready(self):
    while True:
      drdy = self.read(self.LSM6DS3_GYRO_I2C_REG_STAT_REG, 1)[0]
      if drdy & self.LSM6DS3_GYRO_DRDY_GDA:
        break

  def _read_and_avg_data(self) -> list[float]:
    out_buf = [0.0, 0.0, 0.0]
    for _ in range(5):
      self._wait_for_data_ready()
      b = self.read(self.LSM6DS3_GYRO_I2C_REG_OUTX_L_G, 6)
      for j in range(3):
        val = self.parse_16bit(b[j*2], b[j*2+1]) * 70.0  # mdps/LSB for 2000 dps
        out_buf[j] += val
    return [x / 5.0 for x in out_buf]

  def self_test(self, test_type: int):
    # Set ODR to 208Hz, FS to 2000dps
    self.write(self.LSM6DS3_GYRO_I2C_REG_CTRL2_G, self.LSM6DS3_GYRO_ODR_208HZ | self.LSM6DS3_GYRO_FS_2000dps)

    # Wait for stable output
    time.sleep(0.15)
    self._wait_for_data_ready()
    val_st_off = self._read_and_avg_data()

    # Enable self-test
    self.write(self.LSM6DS3_GYRO_I2C_REG_CTRL5_C, test_type)

    # Wait for stable output
    time.sleep(0.05)
    self._wait_for_data_ready()
    val_st_on = self._read_and_avg_data()

    # Disable sensor and self-test
    self.write(self.LSM6DS3_GYRO_I2C_REG_CTRL2_G, 0)
    self.write(self.LSM6DS3_GYRO_I2C_REG_CTRL5_C, 0)

    # Calculate differences and check limits
    test_val = [abs(on - off) for on, off in zip(val_st_on, val_st_off, strict=False)]
    for val in test_val:
      if val < self.LSM6DS3_GYRO_MIN_ST_LIMIT_mdps or val > self.LSM6DS3_GYRO_MAX_ST_LIMIT_mdps:
        raise Exception(f"Gyroscope self-test failed for test type {test_type}")

if __name__ == "__main__":
  s = LSM6DS3_Gyro(1)
  s.init()
  time.sleep(0.1)
  print(s.get_event(0))
  s.shutdown()
